pkg/apis/resource/validation/validation.go

```diff
@@ -17,172 +17,260 @@ limitations under the License.
-	if ref.APIGroup != "" {
-		for _, msg := range utilvalidation.IsDNS1123Subdomain(ref.APIGroup) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("apiGroup"), ref.APIGroup, msg))
-		}
+func validateResourceClaimSpec(spec *resource.ResourceClaimSpec, fldPath *field.Path, stored bool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateDeviceClaim(&spec.Devices, fldPath.Child("devices"), stored)...)
+	if spec.Controller != "" {
+		allErrs = append(allErrs, validateDriverName(spec.Controller, fldPath.Child("controller"))...)
 	}
+	return allErrs
+}
 
-	if ref.Kind == "" {
-		allErrs = append(allErrs, field.Required(fldPath.Child("kind"), ""))
+func validateDeviceClaim(deviceClaim *resource.DeviceClaim, fldPath *field.Path, stored bool) field.ErrorList {
+	allErrs := field.ErrorList{}
+	requestNames := gatherRequestNames(deviceClaim)
+	allErrs = append(allErrs, validateSet(deviceClaim.Requests, resource.DeviceRequestsMaxSize,
+		func(request resource.DeviceRequest, fldPath *field.Path) field.ErrorList {
+			return validateDeviceRequest(request, fldPath, stored)
+		},
+		func(request resource.DeviceRequest) (string, string) {
+			return request.Name, "name"
+		},
+		fldPath.Child("requests"))...)
+	allErrs = append(allErrs, validateSlice(deviceClaim.Constraints, resource.DeviceConstraintsMaxSize,
+		func(constraint resource.DeviceConstraint, fldPath *field.Path) field.ErrorList {
+			return validateDeviceConstraint(constraint, fldPath, requestNames)
+		}, fldPath.Child("constraints"))...)
+	allErrs = append(allErrs, validateSlice(deviceClaim.Config, resource.DeviceConfigMaxSize,
+		func(config resource.DeviceClaimConfiguration, fldPath *field.Path) field.ErrorList {
+			return validateDeviceClaimConfiguration(config, fldPath, requestNames)
+		}, fldPath.Child("config"))...)
+	return allErrs
+}
+
+func gatherRequestNames(deviceClaim *resource.DeviceClaim) sets.Set[string] {
+	requestNames := sets.New[string]()
+	for _, request := range deviceClaim.Requests {
+		requestNames.Insert(request.Name)
+	}
+	return requestNames
+}
+
+func validateDeviceRequest(request resource.DeviceRequest, fldPath *field.Path, stored bool) field.ErrorList {
+	allErrs := validateRequestName(request.Name, fldPath.Child("name"))
+	if request.DeviceClassName == "" {
+		allErrs = append(allErrs, field.Required(fldPath.Child("deviceClassName"), ""))
 	} else {
-		for _, msg := range pathvalidation.IsValidPathSegmentName(ref.Kind) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("kind"), ref.Kind, msg))
+		allErrs = append(allErrs, validateDeviceClassName(request.DeviceClassName, fldPath.Child("deviceClassName"))...)
+	}
+	allErrs = append(allErrs, validateSlice(request.Selectors, resource.DeviceSelectorsMaxSize,
+		func(selector resource.DeviceSelector, fldPath *field.Path) field.ErrorList {
+			return validateSelector(selector, fldPath, stored)
+		},
+		fldPath.Child("selectors"))...)
+	switch request.AllocationMode {
```

Comment:
not so important with 2 cases, but it might be easier to maintain this as something like:

```
if request.AllocationMode == resource.DeviceAllocationModeAll {
  // no dependent fields to validate
}
if request.AllocationMode == resource.DeviceAllocationModeExactCount {
    if request.Count <= 0 {
        // must be GTZ
    }
} else {
    if request.Count != 0 {
        // must not be specified when mode is not Exact
    }
}
if request.AllocationMode == resource.DeviceAllocationModeRange {
    if request.Range == nil {
        // required when mode is Range
    } else {
        // validate range
    }
} else {
    if request.Range != nil {
        // must not be specified when mode is not Range
}
```

Reply:
OK to close this with no action

Reply:
Let's consider that when we get there. Resolving...